tools/setup.py

```diff
@@ -44,29 +45,68 @@ def write_lastchange():
     # ])
 
 
-def get_gn_args():
+# If this text is found in args.gn, we assume it hasn't been hand edited.
+gn_args_header = [
+    "# This file is automatically generated by tools/build.py.",
+    "# REMOVE THIS LINE to preserve any changes you make.", ""
+]
+
+
+def gn_args_are_generated(lines):
+    for line in lines:
+        if re.match("^\s*#.*REMOVE THIS LINE", line):
+            return True
+    return False
+
+
+def read_gn_args(args_filename):
+    if not os.path.exists(args_filename):
+        return (None, False)  # No content, not hand edited.
+
+    with open(args_filename) as f:
+        lines = f.read().splitlines()
+        args = [l.strip() for l in lines if not re.match("^\s*(#|$)", l)]
+        hand_edited = not gn_args_are_generated(lines)
+        return (args, hand_edited)
+
+
+def write_gn_args(args_filename, args):
+    assert not gn_args_are_generated(args)  # No header -> hand crafted.
+    lines = gn_args_header + args
+    assert gn_args_are_generated(lines)  # With header -> generated.
+
+    # Ensure the directory where args.gn goes exists.
+    dir = os.path.dirname(args_filename)
+    if not os.path.isdir(dir):
+        os.makedirs(dir)
+
+    with open(args_filename, "w") as f:
+        f.writelines("\n".join(lines) + "\n")
+
```

Comment:
I feel like this gn args writing is sufficiently complex as to warrant a unit test. Consider moving these functions to util.py and making a test in util_test.py

Reply:
Done.